Add README documentation for @fluidframework/test-version-utils (#9578)
microsoft/FluidFramework: packages/test/test-version-utils/README.md

# @fluidframework/test-version-utils

This is a package for writing and setting up Fluid end to end tests using `mocha` that will generate variants with
a specific driver and different version combinations of Fluid API between layers via `TestObjectProvider` provided
to the test. The different layers are loader, driver, container runtime and data runtime (includes DDS).  Version

> combinations and driver selection can be controlled via the command line for test set up using the provided
>  `describe*` functions.  For advanced usage, test suite can bypass the setup and command line and directly call our

Reviewer: ```suggestion
combinations and driver selection can be controlled via the `mocha` command line, assuming your test uses the provided
 `describe*` functions.  For advanced usage, a test can bypass this mechanism and directly call our
```

Reviewer: Minor rewording, trying to be a little more explicit up front in "how" it works, everything made sense as I read it but I got to the end and didn't really get how you'd use this stuff.

| Loader              |  old   |        |                   |              |
| Driver              |        |  old   |                   |              |
| ContainerRuntime    |        |        | old               |              |
| DataRuntime         |        |        |                   | old          |
| NewLoader           |        |  old   | old               | old          |
| NewDriver           |  old   |        | old               | old          |
| NewContainerRuntime |  old   |  old   |                   | old          |
| NewDataRuntime      |  old   |  old   | old               |              |

### Mocha test setup with layer version combinations

There are three compat `describe*` to generate different combinations, depending of the need of the tests

`describeFullCompat`: generate test variants with compat combinations that varies the version for all layers.

- Used for tests that exercise all layers and will benefits compat combinations of all layers.

`describeLoaderCompat`: generate test variants with compat combinations that only varies the loader version.

- Use for tests that targets the loader layer, and don't care about compat combinations of other layers.

> - Test combination generated: [CompatKind.None, CompatKind.Loader]

Reviewer: Why not also `NewLoader`?

Author: I see that currently it has limited use in examples that doesn't use the `TestObjectProvider` to take advantage of the legacy runtime anyway.   So it doesn't have any benefits with `NewLoader`, as the runtime will always be "new"

Reviewer: Gotcha.  I opened #9581, seems like we should be testing new loader if we're ever targeting loader compatibility specifically.

`describeNoCompat` - generate one test variant that doesn't varies version of any layers.

- Use for tests that doesn't benefit or require any compat testing.
- Test combination generated: [CompatKind.None]

These compat `describe*` functions will also load the APIs with appropriate version and provide the test with a

> `TestObjectProvider` object, where the test can use to access Fluid functionality.  Even compat testing not
> necessary, `TestObjectProvider` provide convenience function and tracking capability, and it allows the test

Reviewer: ```suggestion
`TestObjectProvider` object, where the test can use to access Fluid functionality.  Even when compat testing
is not necessary, `TestObjectProvider` provides convenience function and tracking capability, and it allows the test
```

Can you reword "convenience function and tracking capability"? Not sure what it's about.

Author: will think about the reword.

to enable compat testing easily in the future just by changing the `describe*`.

### Legacy version defaults and installation

By default, N-1, N-2, and LTS (hard coded) test variants are generated.  The versions can be specified using command
line (see below) to run the test against any two versions. This package includes a `mocha` global hook that will
install legacy packages at the beginning of the package based on the `compatVersion` settings.

## Command line options

Tests using the compat `describe*` will be controllable using the command line options when running mocha on
driver selection, versions for compat testing, and compat kind combinations.

```text

> --compatKind <CompatKind> - filter to the compat variant. See above table. Can specify multiple times.

Reviewer: ```suggestion
--compatKind <CompatKind> - filter to the compat variant. See above table. Can specify multiple kinds separated by comma.
```

Author: I think the flag can specify multiple time as well, not just one flag and compatkind in comma.

Reviewer: Like `mocha --compatKind None --compatKind Loader`?  Maybe say `Can be specified multiple times` then

                                Default: undefined (no filter)

> --compatVersion <version> - specify the old version. Relative (to base) or specific version. Can specify multiple times.

Reviewer: ```suggestion
--compatVersion <version> - specify the old version. Relative (to base) or specific version. Can specify multiple versions.
```

                                Default: -1, -2, LTS (hard coded in src/compatConfig.ts)
--baseVersion <version>   - specify the base (new) version. Allow the test to run against any version combinations.
                                Default: <current> (same version of the test package)
--driver <driverType>     - <driverType> = "tinylicious" | "t9s" | "routerlicious" | "r11s" | "odsp" | "local"
                                Default: "local"
--r11sEndpointName <name> - Determine the environment variable name to look for r11s service information to run against
                                Default: "r11s"

> --tenantIndex <number>    - The index into the tenant list to look for odsp server information to run against

Reviewer: Talk about how many there are?  Does it loop modulo that number or do you need to stay within those bounds?

Author: Yep. Modulo. Added verbiage

                                Default: 0
--reinstall               - Force reinstall any required versions.  Default: reuse if a version is already installed.
```

`baseVersion` is a semver or a semver range.

- If it is a semver range, the latest in that range will be picked.

`compatVersion` can be a semver or semver range or an integer apply to the minor version relative to `baseVersion`

- If it is a semver range, the latest in that range will be picked.
- If it is an integer, the value will be apply to the minor version of the `baseVersion`, and create a `^` range
including prerelease versions. The latest in that range will be picked.

>   - i.e. if `baseVersion` is `1.2.3`, and `compatVersion` is `-1`, the resulting range

Reviewer: I thought ^1.1.3-0 would match anything up to 2.0, but that doesn't sound like what we want... Am I confused?

Reviewer: I would expect it to be ~1.1.0 which I think will accept any 1.1.x.

Author: Yeah. I don't want to change the code for now, and it is not a problem right now with 0.x.  Probably need to fix this when we start using the major version.

Reviewer: Then I would update the example here to be 0.x based, like the other example I saw elsewhere.

will be `^1.1.3-0`

We also accept some of the flags via environment variables.

| Environment Variables         | Command line options |
| ----------------------------- | -------------------- |
| fluid__test__compatKind       | --compatKind         |
| fluid__test__compatVersion    | --compatVersion      |
| fluid__test__driver           | --driver             |
| fluid__test__r11sEndpointName | --r11sEndpointName   |
| fluid__test__baseVersion      | --baseVersion        |

> ## Advanced usage

Reviewer: I think a short "sample usage" code block like this would be a great supplement to the words you wrote.

```
import { ensurePackageInstalled, getVersionedTestObjectProvider } from ....

await ensurePackageInstalled("0.50.0");

// get whole test provider
await getVersionedTestObjectProvider("0.50.0");

// get just the ContainerRuntime APIs
await getContainerRuntimeApi("0.50.0");
```

And doc comments on the functions you mention here would be super awesome.

Author: I am trying to not dwell on/expand this part yet, as it is advanced usage, and it hasn't been proven out fully.

> This bypass any configuration of version and direct access to the versions.

Reviewer: ```suggestion
This bypasses any configuration of version used by the describe* functions and provides direct access to the versioned APIs.
```
